typegen/strats/inline.py

```diff
@@ -0,0 +1,230 @@
+import ast
+import functools
+import logging
+import operator
+import pathlib
+import itertools
+
+from constants import TraceData
+from tracing.trace_data_category import TraceDataCategory
+from typegen.strats.gen import TypeHintGenerator
+
+import pandas as pd
+
+logger = logging.getLogger(__name__)
+
+
+class TypeHintTransformer(ast.NodeTransformer):
+    def __init__(self, relevant: pd.DataFrame) -> None:
+        super().__init__()
+        self.df = relevant
+
+    def generic_visit(self, node: ast.AST) -> ast.AST:
+        # Track ClassDefs and FunctionDefs to disambiguate
+        # functions from methods
+        if isinstance(node, ast.ClassDef):
+            for child in filter(lambda c: isinstance(c, ast.FunctionDef), node.body):
+                child.parent = node  # type: ignore
+
+        # Track assignments from Functions
+        if isinstance(node, ast.FunctionDef):
+            for child in filter(
+                lambda c: isinstance(c, ast.AugAssign | ast.AnnAssign | ast.Assign),
+                node.body,
+            ):
+                child.parent = node  # type: ignore
+
+        return super().generic_visit(node)
+
+    def visit_FunctionDef(self, fdef: ast.FunctionDef):
+        logger.debug(f"Applying hints to '{fdef.name}'")
+        # NOTE: disregards *args (fdef.vararg) and **kwargs (fdef.kwarg)
+
+        # parameters
+        param_masks = [
+            self.df[TraceData.CATEGORY] == TraceDataCategory.FUNCTION_ARGUMENT,
+            self.df[TraceData.FUNCNAME] == fdef.name,
+        ]
+        params = self.df[functools.reduce(operator.and_, param_masks)]
+
+        for arg in itertools.chain(fdef.args.posonlyargs, fdef.args.args):
```

Comment:
Put the handling of arguments and returns in different functions for better readability. for example: def add_type_hints_to_function_arguments & add_type_hints_to_function_return.

Reply:
Done